File: frontend/src/api/adminViolations.ts
// 行政端違規勸導管理 API 層(權限鍵 aviol)。
// 伺服器端分頁:漏斗的多選值直接送後端(status/item/filler_id 皆收多值,item 多值=命中任一項),
// 篩選選項走 /admin/violations/options —— 取自實際開立過的紀錄,不是當前這一頁。
// 銷案期限/逾期截止皆為後端推導欄位(resolve_deadline / resolve_expired),前端不再自行推導。
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import dayjs from 'dayjs'
import { api, apiPaged, qs } from './client'

export type ViolationStatusKey = 'violation_open' | 'violation_resolved'
export type ViolationStatus = 'open' | 'resolved'

export interface AdminViolation {
  id: number
  club: string
  date: string // 開立日 YYYY/MM/DD
  location: string
  items: string[]
  other?: string // 其他項目說明
  filler: string // 填寫人姓名
  status: ViolationStatusKey
  resolveNote?: string
  deadline: string // 銷案期限 YYYY/MM/DD(開立日 +1 個月;後端推導)
  expired: boolean // 已逾銷案期限(後端推導;逾期停用銷案)
}

interface AdminViolationOut {
  id: number
  club_name: string
  occurred_on: string
  location: string
  items: string[]
  other: string | null
  filler_name: string
  status: ViolationStatus
  resolve_note: string | null
  resolve_deadline: string | null
  resolve_expired: boolean
}

const toViolation = (v: AdminViolationOut): AdminViolation => ({
  id: v.id,
  club: v.club_name,
  date: dayjs(v.occurred_on).format('YYYY/MM/DD'),
  location: v.location,
  items: v.items,
  other: v.other ?? undefined,
  filler: v.filler_name,
  status: v.status === 'open' ? 'violation_open' : 'violation_resolved',
  resolveNote: v.resolve_note ?? undefined,
  deadline: v.resolve_deadline ? dayjs(v.resolve_deadline).format('YYYY/MM/DD') : '',
  expired: v.resolve_expired,
})

export interface AdminViolationListParams {
  items?: string[]
  fillerIds?: number[]
  statuses?: ViolationStatus[]
  /** 逾期篩選(後端僅對未銷案生效):true=已截止、false=未逾期 */
  expired?: boolean
  /** 排序白名單:date/location/items/filler/deadline/status;前綴 - 為降冪 */
  sort?: string
  page: number
  pageSize: number
}

export interface ViolationOptions {
  items: string[]
  fillers: { id: number; name: string }[]
}

export const ALL_VIOLATION_STATUSES: ViolationStatus[] = ['open', 'resolved']
export const VIOLATION_STATUS_LABEL: Record<ViolationStatus, string> = {
  open: '未銷案',
  resolved: '已銷案',
}
export const DEADLINE_LABELS = ['未逾期', '已截止']

/**
 * 兩個漏斗(狀態、銷案期限)的選取值 → 查詢參數。
 *
 * 期限漏斗只對未銷案有意義(已銷案該欄顯示「—」,兩個選項都不吻合):只選一邊 →
 * expired 布林(後端一併限未銷案);兩邊都選 → 等於「僅未銷案」。兩個漏斗取交集,
 * 交集為空(如「已銷案」+ 期限兩項全選)是合法選擇,結果就是沒有列 —— 呼叫端據此不發查詢,
 * 不可退回「不帶 status」的全部。填寫人同理:選了名字卻對不到 id 一律強制空集。
 */
export function violationFilterParams(sel: {
  statusLabels: string[]
  deadlineLabels: string[]
  fillerNames: string[]
  fillers: { id: number; name: string }[]
}): { statuses: ViolationStatus[]; expired?: boolean; fillerIds?: number[] } {
  const byStatus = sel.statusLabels.length
    ? ALL_VIOLATION_STATUSES.filter((s) => sel.statusLabels.includes(VIOLATION_STATUS_LABEL[s]))
    : ALL_VIOLATION_STATUSES
  const byDeadline: ViolationStatus[] =
    sel.deadlineLabels.length === 2 ? ['open'] : ALL_VIOLATION_STATUSES
  const fillerIds = sel.fillerNames.length
    ? sel.fillers.filter((f) => sel.fillerNames.includes(f.name)).map((f) => f.id)
    : undefined
  return {
    statuses: byStatus.filter((s) => byDeadline.includes(s)),
    expired: sel.deadlineLabels.length === 1 ? sel.deadlineLabels[0] === '已截止' : undefined,
    fillerIds: fillerIds && fillerIds.length === 0 ? [-1] : fillerIds,
  }
}

const keys = {
  all: ['adminViolations'] as const,
  list: (p: AdminViolationListParams) => ['adminViolations', 'list', p] as const,
  options: ['adminViolations', 'options'] as const,
}

export function useAdminViolations(p: AdminViolationListParams, opts: { enabled?: boolean } = {}) {
  return useQuery({
    queryKey: keys.list(p),
    enabled: opts.enabled ?? true,
    queryFn: () =>
      apiPaged<AdminViolationOut[]>(
        `/admin/violations${qs({
          status: p.statuses,
          item: p.items,
          filler_id: p.fillerIds?.map(String),
          expired: p.expired,
          sort: p.sort,
          page: p.page,
          page_size: p.pageSize,
        })}`,
      ).then(({ data, total }) => ({ rows: data.map(toViolation), total })),
    // 不留上一份:query key 含篩選條件,沿用舊資料等於把別的篩選結果當成這次的
  })
}

export function useViolationOptions() {
  return useQuery({
    queryKey: keys.options,
    queryFn: () => api<ViolationOptions>('/admin/violations/options'),
  })
}

export function useResolveViolation() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ id, note }: { id: number; note: string }) =>
      api<AdminViolationOut>(`/admin/violations/${id}/resolve`, {
        method: 'POST',
        body: JSON.stringify({ note }),
      }),
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: keys.all })
      void qc.invalidateQueries({ queryKey: ['adminOverview'] }) // 未銷案違規數字卡
    },
  })
}

File: backend/app/api/v1/admin_violations.py
"""行政端:違規勸導管理(開立日 +1 個月銷案期限,逾期截止)。

列表支援排序(日期/地點/項目/填寫人/期限/狀態白名單)與過濾;
預設排序=未銷案在前、時間升冪;逾期後銷案 API 拒絕(RESOLVE_EXPIRED)。
"""

from datetime import date
from typing import Annotated

import sqlalchemy as sa
from fastapi import APIRouter, BackgroundTasks, Depends, Query, Request

from app.api.pagination import Pagination, parse_sort
from app.core.deps import CurrentUser, DbDep, client_ip, require_permission
from app.core.errors import conflict, not_found
from app.models import Club, User, Violation
from app.models.enums import ViolationStatus
from app.schemas.admin import (
    AdminViolationOut,
    ResolveViolationIn,
    ViolationFillerOut,
    ViolationOptionsOut,
)
from app.schemas.common import ApiResponse
from app.services import audit, notify, violation_service

router = APIRouter(prefix="/admin/violations", tags=["admin"])

ViolationAdmin = Annotated[CurrentUser, Depends(require_permission("aviol"))]

_FILLER = sa.orm.aliased(User)

# 未銷案在前(業務語意),不是列舉字面值 —— 目前 'open' < 'resolved' 只是巧合
_STATUS_ORDER = sa.case((Violation.status == ViolationStatus.OPEN, 0), else_=1)

# 排序白名單:期限=開立日+1 個月(單調),以開立日排序等價
_SORTABLE = {
    "date": Violation.occurred_on,
    "location": Violation.location,
    "items": sa.func.array_to_string(Violation.items, "、"),
    "filler": _FILLER.name,
    "deadline": Violation.occurred_on,
    "status": _STATUS_ORDER,
    "created_at": Violation.created_at,
}


def _to_out(v: Violation, club_name: str, filler_name: str, today: date) -> AdminViolationOut:
    out = AdminViolationOut.model_validate(v)
    out.club_name = club_name
    out.filler_name = filler_name
    out.resolve_deadline = violation_service.resolve_deadline(v)
    out.resolve_expired = violation_service.resolve_expired(v, today)
    return out


@router.get("/options")
async def violation_options(user: ViolationAdmin, db: DbDep) -> ApiResponse[ViolationOptionsOut]:
    """篩選選項取自實際開立過的紀錄:承辦不必先翻到那一頁才篩得到,目錄改過也不漏舊項目。"""
    items_sub = sa.select(sa.func.unnest(Violation.items).label("item")).subquery()
    items = await db.scalars(sa.select(items_sub.c.item).distinct().order_by(items_sub.c.item))
    fillers = await db.execute(
        sa.select(_FILLER.id, _FILLER.name)
        .join(Violation, Violation.filler_id == _FILLER.id)
        .distinct()
        .order_by(_FILLER.name)
    )
    return ApiResponse(
        data=ViolationOptionsOut(
            items=list(items),
            fillers=[ViolationFillerOut(id=i, name=n) for i, n in fillers],
        )
    )


@router.get("")
async def list_violations(
    user: ViolationAdmin,
    db: DbDep,
    page: Pagination,
    sort: str | None = None,
    status: Annotated[list[ViolationStatus] | None, Query()] = None,
    club_id: Annotated[list[int] | None, Query()] = None,
    filler_id: Annotated[list[int] | None, Query()] = None,
    item: Annotated[list[str] | None, Query()] = None,
    location: str | None = Query(None),
    date_from: date | None = None,
    date_to: date | None = None,
    expired: bool | None = Query(None),
) -> ApiResponse[list[AdminViolationOut]]:
    # status/club_id/filler_id/item 可重複帶多值(畫面的漏斗是多選);item 多值=命中任一項
    query = (
        sa.select(Violation, Club.name, _FILLER.name)
        .join(Club, Violation.club_id == Club.id)
        .join(_FILLER, Violation.filler_id == _FILLER.id)
    )
    if status:
        query = query.where(Violation.status.in_(status))
    if club_id:
        query = query.where(Violation.club_id.in_(club_id))
    if filler_id:
        query = query.where(Violation.filler_id.in_(filler_id))
    if item:
        query = query.where(Violation.items.overlap(item))
    if location:
        query = query.where(Violation.location.ilike(f"%{location}%"))
    if date_from:
        query = query.where(Violation.occurred_on >= date_from)
    if date_to:
        query = query.where(Violation.occurred_on <= date_to)
    today = violation_service.today_taipei()
    if expired is not None:
        # 期限篩選僅對未銷案有意義(已銷案顯示 —)
        query = query.where(Violation.status == ViolationStatus.OPEN)
        deadline = violation_service.deadline_sql()
        query = query.where(deadline < today if expired else deadline >= today)

    if sort:
        query = query.order_by(*parse_sort(sort, _SORTABLE, None), Violation.id)
    else:
        # 預設排序:未銷案在前,各組內時間升冪
        query = query.order_by(_STATUS_ORDER, Violation.occurred_on.asc(), Violation.id)

    total = await db.scalar(sa.select(sa.func.count()).select_from(query.subquery()))
    rows = await db.execute(query.offset(page.offset).limit(page.page_size))
    data = [_to_out(v, club_name, filler_name, today) for v, club_name, filler_name in rows]
    return ApiResponse(data=data, meta=page.meta(total or 0))


@router.post("/{violation_id}/resolve")
async def resolve_violation(
    violation_id: int,
    body: ResolveViolationIn,
    user: ViolationAdmin,
    db: DbDep,
    request: Request,
    background: BackgroundTasks,
) -> ApiResponse[AdminViolationOut]:
    violation = await db.scalar(
        sa.select(Violation).where(Violation.id == violation_id).with_for_update()
    )
    if violation is None:
        raise not_found("找不到違規勸導紀錄")
    if violation.status != ViolationStatus.OPEN:
        raise conflict("此紀錄已銷案")
    today = violation_service.today_taipei()
    if violation_service.resolve_expired(violation, today):
        # 開立日 +1 個月逾期即截止:不再受理銷案,−1 扣分成立
        raise conflict("已逾銷案期限,不再受理銷案", code="RESOLVE_EXPIRED")

    violation.status = ViolationStatus.RESOLVED
    violation.resolve_note = body.note
    audit.record(
        db,
        action="violation_resolved",
        user=user,
        detail=f"violation={violation.id}",
        ip=client_ip(request),
    )
    await db.commit()

    club = await db.get(Club, violation.club_id)
    background.add_task(
        notify.club_event,
        "approve",
        "違規勸導已銷案",
        f"{club.name}:{violation.occurred_on} {violation.location}",
        club.discord_webhook_url,
    )
    filler = await db.get(User, violation.filler_id)
    return ApiResponse(data=_to_out(violation, club.name, filler.name if filler else "", today))
